### ISSUE TITLE
[Question] Reward Shaping

### ISSUE THREAD

**Issue by scrunguss** [OliverOverend/gym-simplifiedtetris#26]
Hi,

What is the reasoning behind the calculation of the potential-based rewards function based on the holes feature? I'm referring to what is happening in the _get_reward() function in _potential_based_shaping_reward.py. Thanks

**Comment by OliverOverend** [comment]
Hi,

An agent selecting actions uniformly at random rarely receives a positive reward when playing the standard game of Tetris. In this scenario, many episodes are required before the agent begins to learn anything. In other domains, a precarious means of avoiding this problem is to inject human knowledge into the reward function via _reward shaping_.

The idea behind this is to create a _shaping reward_, R' = R + F, where R is the original reward function, and F is the _shaping reward function_ ([link](https://www.semanticscholar.org/paper/Policy-Invariance-Under-Reward-Transformations%3A-and-Ng-Harada/94066dc12fe31e96af7557838159bde598cb4f10)). Ideally the shaping reward should have _policy invariance_, which is achieved if, and only if, the shaping reward function is _potential-based_ (i.e., a difference in potential functions). 

I chose the potential function to be a function of the well-known _holes_ feature because the number of holes in a given state is (loosely speaking) inversely proportional to the potential of a state. As for L31–38, I wanted the difference in potentials to be in [-1, 1] to improve the stability of neural network convergence. I also wanted the agent to frequently receive non-zero rewards (since bad-performing agents in the standard game of Tetris rarely receive non-zero rewards). Hence, the value of _holes_ was scaled by using the smallest and largest values of _holes_ seen thus far to obtain a value in [0, 1). The result of this was then subtracted from 1 (to obtain a value in (0, 1]) because a state with a larger value of _holes_ has a smaller potential (generally speaking). The function `numpy.clip` is redundant here. 

https://github.com/OliverOverend/gym-simplifiedtetris/blob/832a0e99b52ec0c13bad1badc0dc1ba6453a6981/gym_simplifiedtetris/envs/reward_shaping/_potential_based_shaping_reward.py#L31-L38

L38 calculates the shaping reward R'. Notice that gamma was set to 1, which isn't strictly allowed since it should be less than 1 according to Theorem 1 in [this](https://www.semanticscholar.org/paper/Policy-Invariance-Under-Reward-Transformations%3A-and-Ng-Harada/94066dc12fe31e96af7557838159bde598cb4f10) paper. I found that the agent rarely received positive rewards using this reward function because the agent was frequently transitioning to states with a lower potential (since it was rarely clearing lines).
